src/main/java/org/kohsuke/github/GitHubClient.java

```diff
@@ -111,33 +93,45 @@
         this.apiUrl = apiUrl;
         this.connector = connector;
 
-        if (oauthAccessToken != null) {
-            encodedAuthorization = "token " + oauthAccessToken;
+        // Prefer credential configuration via provider
+        if (credentialProvider != null) {
+            this.credentialProvider = credentialProvider;
         } else {
-            if (jwtToken != null) {
-                encodedAuthorization = "Bearer " + jwtToken;
-            } else if (password != null) {
-                String authorization = (login + ':' + password);
-                String charsetName = StandardCharsets.UTF_8.name();
-                encodedAuthorization = "Basic "
-                        + Base64.getEncoder().encodeToString(authorization.getBytes(charsetName));
-            } else {// anonymous access
-                encodedAuthorization = null;
+            if (oauthAccessToken != null) {
+                this.credentialProvider = ImmutableCredentialProvider.fromOauthToken(oauthAccessToken);
+            } else {
+                if (jwtToken != null) {
+                    this.credentialProvider = ImmutableCredentialProvider.fromJwtToken(jwtToken);
+                } else if (password != null) {
+                    this.credentialProvider = ImmutableCredentialProvider.fromLoginAndPassword(login, password);
+                } else {// anonymous access
+                    this.credentialProvider = CredentialProvider.ANONYMOUS;
+                }
             }
         }
 
         this.rateLimitHandler = rateLimitHandler;
         this.abuseLimitHandler = abuseLimitHandler;
         this.rateLimitChecker = rateLimitChecker;
 
-        if (login == null && encodedAuthorization != null && jwtToken == null) {
-            GHMyself myself = fetch(GHMyself.class, "/user");
-            login = myself.getLogin();
-            if (myselfConsumer != null) {
-                myselfConsumer.accept(myself);
+        this.login = getCurrentUser(login, jwtToken, myselfConsumer);
+    }
+
+    @Nullable
+    private String getCurrentUser(String login, String jwtToken, Consumer<GHMyself> myselfConsumer) throws IOException {
+        if (login == null && this.credentialProvider.getEncodedAuthorization() != null && jwtToken == null) {
+            try {
+                GHMyself myself = fetch(GHMyself.class, "/user");
+                if (myselfConsumer != null) {
+                    myselfConsumer.accept(myself);
+                }
+                return myself.getLogin();
+            } catch (IOException e) {
```

Comment:
Why add this `try-catch`?  It changes the user visible behavior.

Reply:
If we use a credentialProvider, we no longer know what kind of credentials we have here. Some credentials don't support checking who is making the call (api call to "/user"), and will fail, so the only way that I see to check if the credentials being used support "user lookup", is to actually make the call and handle the case where the call fails (e.g: installation tokens).

Reply:
> If we use a credentialProvider, we no longer know what kind of credentials we have here. 

What about having a method on `CredentialsProvider`?  `getLogin(Consumer<GHMyself> myselfConsumer)` or something similar?  Then `getCurrentUser()` wouldn't check `jwtToken` or `login` or whatever, the credential itself would decide what to do.  `getLogin()` on `JWTTokenProvider` would always return `null`, while for `fromOauthToken` it would have code from lines 124-128.

Reply:
Hmm, or maybe a `CredentialsProvider` would have `boolean requiresMyself()` to control whether this code is called and then `getLogin(@CheckForNull GHMyself)` that is always called, `return credentialProvider.getLogin(myself);`.